### PR TITLE
Producer: add listener for bytes written to topic

### PR DESCRIPTION
**Pull request by vikmik**
This change allows us to record the number of bytes written to a topic
This PR adds a generic listener in the form of a `kgo.HookProduceBatchWritten`. The code that creates the producer may pass an optional listener function to be notified of writes to the topic.

### CODE REVIEW

**Comment by vikmik** [comment]
Looks like the tests are already failing in `main`. Not sure what this is about yet

File: kafka/producer_test.go [hunk 1]
```diff
@@ -146,7 +154,8 @@ func TestNewProducerBasic(t *testing.T) {
 			var actual []apmqueue.Record
 			orderingKeyToPartitionM := make(map[string]int32)
 			client.AddConsumeTopics(namespacedTopic)
-			for i := 0; i < len(batch)*partitionCount && len(actual) < len(batch); i++ {
```
Comment (vikmik) [inline kafka/producer_test.go:149]: Unless I'm missing something, that first condition seemed incorrect ; if we want to read all records we must ensure that `len(actual) < len(batch)`.

File: kafka/producer.go [hunk 2]
```diff
@@ -81,8 +81,20 @@ type ProducerConfig struct {
 	// If $KAFKA_PRODUCER_COMPRESSION_CODEC is not specified, then
 	// the default behaviour of franz-go is to use [snappy, none].
 	CompressionCodec []CompressionCodec
+
+	// BatchWriteListener specifies a callback function that will notify the caller of writes.
```
Comment (marclop) [inline kafka/producer.go:85]:
```suggestion
	// BatchWriteListener is called per topic/partition after a batch is
	// successfully produced to a Kafka broker.
```

File: kafka/producer.go [hunk 3]
```diff
@@ -81,8 +81,20 @@ type ProducerConfig struct {
...
+	BatchListener BatchWriteListener
+}
+
+type BatchWriteListener func(topic string, bytesWritten int)
+
+func (l BatchWriteListener) OnProduceBatchWritten(_ kgo.BrokerMetadata,
```
Comment (marclop) [inline kafka/producer.go:91]: Would you mind adding a GoDoc comment for the type and public method?

File: kafka/producer_test.go [hunk 4]
```diff
@@ -198,6 +207,13 @@ func TestNewProducerBasic(t *testing.T) {
 			//lint:ignore SA1012 passing a nil context is a valid use for this call.
 			fetches := client.PollRecords(nil, 1)
 			assert.Len(t, fetches.Records(), 0)
+
+			// Ensure that we recorded bytes written to the topic.
+			mut.Lock()
+			require.Len(t, topicBytesWritten, 1)
+			require.Contains(t, topicBytesWritten, "name_space-default-topic")
+			require.GreaterOrEqual(t, topicBytesWritten["name_space-default-topic"], 100)
```
Comment (marclop) [inline kafka/producer_test.go:215]: Shouldn't this be `.Equal`? In which cases will this be Greater?

File: kafka/producer.go [hunk 5]
```diff
@@ -81,8 +81,20 @@ type ProducerConfig struct {
 	// If $KAFKA_PRODUCER_COMPRESSION_CODEC is not specified, then
 	// the default behaviour of franz-go is to use [snappy, none].
 	CompressionCodec []CompressionCodec
+
+	// BatchWriteListener specifies a callback function that will notify the caller of writes.
+	BatchListener BatchWriteListener
+}
+
+type BatchWriteListener func(topic string, bytesWritten int)
```
Comment (marclop) [inline kafka/producer.go:89]: I like how you encapsulated a subset of the behavior in a new type.

**Review by marclop — COMMENTED** [review]
Change LGTM. Just some stylistic comments and a question on the test assertion.

File: kafka/producer_test.go (hunk 4 above)
Comment (vikmik) [reply kafka/producer_test.go:215]: This is slightly non-deterministic - the number of actual topic writes may vary, and hence the compressed bytes may vary.
I should probably add a comment for this

**Review by marclop — APPROVED** [review]
Thanks for addressing my comments. LGTM!

**Pull request merged**